index.js

```diff
@@ -104,51 +77,279 @@ module.exports = {
+      this.internalState.transporter = transporter;
+      // ensure we can do basic JSON-RPC over this connection
+      this.version(function (errorOrResult) {
+        if (errorOrResult instanceof Error || errorOrResult.error) return initialConnectCallback(errorOrResult);
+        this.setupBlockSubscription(function () { initialConnectCallback(null); });
+      }.bind(this));
+    }.bind(this));
+  },
+
+  /**
+   * Resets the global state of this module to default.
+   */
+  resetState: function() {
+    // stop any pending timers
+    clearInterval((this.internalState || {}).newBlockIntervalTimeoutId);
+
+    // reset configuration to defaults
+    this.configuration = {
+      httpAddresses: [],
+      wsAddresses: [],
+      ipcAddresses: [],
+      connectionTimeout: 3000,
+      errorHandler: null,
+    };
+
+    // redirect any not-yet-received responses to /dev/null
+    var oldMessageHandlerObject = (this.internalState || {}).shimMessageHandlerObject || {};
+    var newMessageHandlerObject = { realMessageHandler: this.blockchainMessageHandler.bind(this) };
+    oldMessageHandlerObject.realMessageHandler = function () {};
+
+    // reset state to defaults
+    this.internalState = {
+      transporter: null,
+      outstandingRequests: {},
+      subscriptions: {},
+      lastSeenLatestBlockNumber: null,
+      lowestSeenSequentialBlock: null,
+      newBlockIntervalTimeoutId: null,
+      shimMessageHandlerObject: newMessageHandlerObject,
+      // by binding this function to `shimMessageHandlerObject`, its `this` value will be a pointer to an object that we can mutate before replacing when reset
+      shimMessageHandler: function (error, jso) { this.realMessageHandler(error, jso); }.bind(newMessageHandlerObject)
+    };
+
+    // reset public state
+    this.block = null;
+    this.excludedFromTxRelay = {};
+    this.gasPrice = 20000000000;
+    this.notifications = {};
+    this.rawTxMaxNonce = -1;
+    this.rawTxs = {};
+    this.requests = 1;
+    this.txs = {};
+  },
+
+  // TODO: add support for caller passing in request and response type information
+  /**
+   * Used internally.  Submits a remote procedure call to the blockchain.
+   *
+   * @param {!object} jso - The JSON-RPC call to make.
+   * @param {?string} transportRequirements - ANY, SYNC or DUPLEX.  Will choose best available transport that meets the requirements.
+   * @param {?function(?Error, ?object):void} callback - Called when a response to the request is received.  May only be null if preferredTransport is SYNC.
+   * @returns {void|?Error|?object} - Returns the error or result if the operation is synchronous.
+   */
+  submitRequestToBlockchain: function (jso, transportRequirements, callback) {
+    var syncErrorOrResult;
+    if (transportRequirements === "SYNC") callback = function (error, result) { return (syncErrorOrResult = (error || result)); };
+
+    // allow people to call this function incorrectly rather than failing fast
```

Comment:
AFAIK this is a well-accepted way to provide a flexible function signature in JS.  The convention that the last argument is the callback (even if you omit some parameters) is used everywhere in the standard library (i.e., Node.js's built-in modules).